### PR TITLE
TEP-0097 breakpoint before steps for taskrun

### PR DESCRIPTION
**Pull request by chengjoey**
# Changes
breakpoint before steps for taskrun
[TEP-0097](https://github.com/tektoncd/community/blob/main/teps/0097-breakpoints-for-taskruns-and-pipelineruns.md#alternatives)
#5691 

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [x] Has [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) if any changes are user facing, including updates to minimum requirements e.g. Kubernetes version bumps
- [x] Has [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [x] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [x] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including functionality, content, code)
- [x] Has a kind label. You can add one by adding a comment on this PR that contains `/kind <type>`. Valid types are bug, cleanup, design, documentation, feature, flake, misc, question, tep
- [x] Release notes block below has been updated with any user facing changes (API changes, bug fixes, changes requiring upgrade notices or deprecation warnings). See some examples of [good release notes](https://github.com/tektoncd/community/blob/main/standards.md#good-release-notes).
- [ ] Release notes contains the string "action required" if the change requires additional action from users switching to the new release

# Release Notes

```release-note
breakpoint before a step
```

### CODE REVIEW

**Comment by chengjoey** [comment]
> @chengjoey @chitrangpatel I think this only needs a second reviewer, correct? If so I'll add it to my review list
right, please review this🙏

**Comment by chitrangpatel** [comment]
@chengjoey can you please rebase when you get a chance?

**Comment by chitrangpatel** [comment]
This needs to be rebased. Moving to the next release to give some more time. Please let us know otherwise.

**Comment by chengjoey** [comment]
> @chengjoey can you please rebase when you get a chance?

done~

**Comment by chengjoey** [comment]
@chitrangpatel there are some code format errors., please lgtm again~

**Comment by chitrangpatel** [comment]
Note that the integration tests are currently failing:
```
+ docker run -d --restart=always -p 5000:5000 --name registry.local registry:2
Unable to find image 'registry:2' locally
docker: Error response from daemon: toomanyrequests: You have reached your pull rate limit. You may increase the limit by authenticating and upgrading: https://www.docker.com/increase-rate-limit.
```
😢 
We need to wait a bit before submitting the tests again

**Comment by pritidesai** [comment]
@chengjoey please fix the linting error.

**Comment by chitrangpatel** [comment]
@chengjoey please rebase. This will pull in the image tagging changes introduced to hopefully get us past the CI Docker rate limit errors

**Comment by chengjoey** [comment]
> @chengjoey please rebase. This will pull in the image tagging changes introduced to hopefully get us past the CI Docker rate limit errors

i have reased, but it doesn't seem to work, I'll try test again later

**Comment by chitrangpatel** [comment]
> > @chengjoey please rebase. This will pull in the image tagging changes introduced to hopefully get us past the CI Docker rate limit errors
> 
> i have reased, but it doesn't seem to work, I'll try test again later

yeah I think we are still being affected by the ongoing rate pull effect which exists for 6 hours!!

**Comment by chengjoey** [comment]
image

It looks a bit strange, is it because of golint upgrade?

**Comment by chengjoey** [comment]
@chitrangpatel , now all the checks have passed~

File: pkg/entrypoint/entrypointer.go [hunk 1]
```diff
@@ -52,6 +52,22 @@ const (
...
+
+// DebugBeforeStepError is an error means mark before step breakpoint failure
+type DebugBeforeStepError string
+
+func (e DebugBeforeStepError) Error() string {
+	return string(e)
+}
+
+var (
+	errDebugBeforeStep = DebugBeforeStepError("before step breakpoint error file, user decided to skip the current step execution")
```
Comment (afrittoli) [inline pkg/entrypoint/entrypointer.go:68]: I'm not sure I understand the meaning of this.
I'm guessing that "an error was encountered while waiting for the before-step breakpoint file because the user decided to skip the execution of the step"? Could you clarify?

Comment (afrittoli) [reply pkg/entrypoint/entrypointer.go:68]: Maybe it could be rephrased as: "before-step breakpoint: step execution skipped by the user"?

File: pkg/entrypoint/entrypointer.go [hunk 2]
```diff
@@ -301,25 +325,7 @@ func (e Entrypointer) Go() error {
...
-		if e.ArtifactsDirectory != "" {
-			artifactsDir = e.ArtifactsDirectory
-		}
-		fp = filepath.Join(artifactsDir, "provenance.json")
-		artifacts, err = readArtifacts(fp, result.TaskRunArtifactsResultType)
-		if err != nil {
-			logger.Fatalf("Error while handling task artifacts: %s", err)
-		}
-		output = append(output, artifacts...)
+		e.appendArtifactOutputs(&output, logger)
```
Comment (afrittoli) [inline pkg/entrypoint/entrypointer.go:328]: What's the reason for moving this code to a function? It does not help with testing and it is not re-used anywhere else.

File: cmd/entrypoint/main.go [hunk 3]
```diff
@@ -210,7 +196,7 @@ func main() {
 			// in both cases has an ExitStatus() method with the
 			// same signature.
 			if status, ok := t.Sys().(syscall.WaitStatus); ok {
-				checkForBreakpointOnFailure(e, breakpointExitPostFile)
+				e.CheckForBreakpointOnFailure()
```
Comment (afrittoli) [inline cmd/entrypoint/main.go:199]: Nice, but it seems unrelated to this PR.
Is the reason for making `CheckForBreakpointOnFailure` public is to provide unit test coverage?

**Review by afrittoli — COMMENTED** [review]
Thank you for this! Just a few minor comments, otherwise it looks good!
Please let me know if you prefer to merge this and address the comments in a follow-up PR.

**Comment by chengjoey** [comment]
> Thank you for this! Just a few minor comments, otherwise it looks good! Please let me know if you prefer to merge this and address the comments in a follow-up PR.

Thanks @afrittoli for your comments. Yes, I hope to merge this PR and fix the issues raised in your comments in a follow-up PR, and I will fix them before the next milestone.

**Comment by chitrangpatel** [comment]
Merging now.
/lgtm
